kpet/cmd_arch.py

```diff
@@ -41,9 +46,16 @@ def main(args):
     database = data.Base(args.db)
     if args.action == 'list':
         regex = re.compile(args.regex or ".*")
-        for arch in sorted(database.arches):
-            # filter using regex; if it's not set, print all
-            if regex.fullmatch(arch):
-                print(arch)
+        tree_regex = re.compile(args.tree or ".*")
+
+        filtered = list(value['arches'] for (key, value)
+                        in database.trees.items()
+                        if tree_regex.fullmatch(key))
+        # Flattens the list of lists and removes duplicates
+        filtered = {y for x in filtered for y in x}
```

Comment:
This is a bit too complicated. I also have no idea how that set
comprehension syntax works :)

I think this would be simpler to read:

    tree_arches = set()
    for name, value in database.trees.items():
        if tree_regex.fullmatch(name):
            tree_arches |= set(value['arches'])

    for arch in sorted(filter(regex.fullmatch, tree_arches)):
        print(arch)

You don't need to explicitly convert the output of `filter` to a list as
`sorted` accepts an iterator.